### ISSUE TITLE
selective_scan_ref deltaB_u doesn't follow Eqn 4 in paper

### ISSUE THREAD

**Issue by hrbigelow** [state-spaces/mamba#19]
Hi,

It seems like the reference implementation doesn't correctly implement equation 4, specifically deltaB.
 While the computation for `deltaA` seems to correspond with the paper eqn 4:

```python
deltaA = torch.exp(torch.einsum('bdl,dn->bdln', delta, A))
```

Eqn 4:

![image](https://github.com/state-spaces/mamba/assets/5355286/81ebe8a8-9452-427e-83b0-d268664b67a1)

The section for `deltaB_u`  in the [code](https://github.com/state-spaces/mamba/blob/6dbfc4553a98c81e1e93b8fd2d5abf387f5c09ee/mamba_ssm/ops/selective_scan_interface.py#L124-L134) is just:

```python
    else:
        if B.dim() == 3:
            deltaB_u = torch.einsum('bdl,bnl,bdl->bdln', delta, B, u)
        else:
            B = repeat(B, "B G N L -> B (G H) N L", H=dim // B.shape[1])
            deltaB_u = torch.einsum('bdl,bdnl,bdl->bdln', delta, B, u)
    if is_variable_C and C.dim() == 4:
        C = repeat(C, "B G N L -> B (G H) N L", H=dim // C.shape[1])
    last_state = None
    for i in range(u.shape[2]):
        x = deltaA[:, :, i] * x + deltaB_u[:, :, i]
```

Maybe I missed something more subtle but I couldn't find it.

Thanks in advance for any help you could provide.  Mostly, I'm just trying to figure out the indexing since the formula in Eqn 4 looks a lot like matmuls but it is not.  So, I am not sure quite what the exponentiation means.

Thanks very much,

Henry

**Comment by albertfgu** [comment]
Yes, we took an approximation that didn't change the performance much but is slightly easier to implement. This is the first order approximation on the discretization coefficient in front of $B$ (i.e. the Euler approximation, but only applied to the $B$ term).

**Comment by xiaoliangbai** [comment]
A related question on A_bar, shouldn't it be matrix exponential e^(Delta A)? The first order approximation would be Eye + Delta .* A; In the reference code, element wise exp function is used:

def selective_scan_ref(...):
      ...
      deltaA = torch.exp(torch.einsum('bdl,dn->bdln', delta, A))
      
Thanks

**Comment by xiaoliangbai** [comment]
I see, the network will learn all the eigen values of A, and eigen vectors absorbed by surrounding projections (weights also learned).
Thanks for the wonderful work!
